design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,127 @@
+ 
+![work flow](./workflow.png)
+
+
+1. The enhanced Rook-Ceph Operator sees an CephObjectBucket in the dev-user namespace.  The operator extracts the CephObjectStoreUser from the CephObjectBucket. From the CephObjectStoreUser the StorageClass, if present, is used to determine the associated object store. The operator uses the name of the CephObjectStoreUser to get the associated Key Pair Secret.  If the StorageClass is not defined in the CephObjectStoreUser then the object store name must be present. Also, in the case of omitting StorageClass, the object user and object store must reside in the same namespace.
+1. With the object store IP, port, and user name from step 1, the operator makes an S3 API call to create a new bucket with the CephObjectStoreUser as the owner.
+1. The operator creates a bucket ConfigMap in the dev-user namespace. The Host, Port, BucketName, and SSL (bool) are stored under data: .  For the purpose of bookkeeping, the namespace/CephObjectStoreUser.metadata.name and namespace/Secret.metadata.name are stored as annotations.
+1. An app Pod may then mount the Secret (annotated in the config map)  and the ConfigMap keys to begin accessing the bucket. 
+
+**Use Case:** Delete an Object Bucket 
+_As a Kubernetes user, I want to use kubectl to delete object bucket API resources owned by a given object user. (Note: the underlying S3 bucket is not deleted. Permanent deletion is left to the admin.)_
+
+1. The user deletes the CephObjectBucket.  This triggers an automated cleanup sequence:
+1. The CephObjectBucket is marked for deletion first and left in the foreground. Deletion is blocked pending the next step.
+1. The respective ConfigMap is deleted.
+1. The CephObjectBucket is garbage collected.
+
+## API Specifications
+
+### CephObjectBucket
+
+```yaml
+apiVersion: object.k8s.io/v1
+kind: CephObjectBucket
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  ownerReferences: # added by provisioner
+  - apiVersion: ceph.rook.io/v1
+    blockOwnerDeletion: true
+    kind: ObjectStoreUser
+    name: my-user-1
+    uid: 1234-3456-7839-1231
+spec:
+  bucketName:<string>         [1]
+  generateBucketName:<string> [1]
+  ownerUser: my-user # required, name of object user’s metadata.Name
+status:
+  phase: [ “creating”, “available”, ...]
+
+```
+
+[1] `bucketName` and `generateBucketName` are mutually exclusive. It is expected that `generateBucketName` be the common usage. This field produces a random bucket name beginning with the value of this stanza. If bucketName is defined then that value alone is used as the name of the bucket. Within an object store, buckets must be unique, so if bucketName is used there may be a higher chance of name collision. The api will continue retrying if the bucket cannot be created.
+
+### CephObjectBucketConfigMap
```

Comment:
what do you think about capturing all connection information, including both location information and credentials information, into a single secret? that may be an easier model for apps/pods to consume everything they need to connect from a single k8s object. thoughts?

Reply:
We've considered adding the endpoint to the secret and eliminating the configMap. A couple of thoughts:
1. the endpoint value doesn't need to be encrypted/hidden so it better "fits" the configMap use case.
1. if I have one user (and thus 1 secret) and I want to create two buckets for the same user, then the secret would need a list of endpoints. But if the real bucket name is randomized then the secret needs to associate the endpoint to the bucket CRD name.

Reply:
@jbw976 Do you mean copying the access keys into a new secret?  It would definitely simplify things but it does mean that changes to the secret (changning credentials) wouldn't propagate to the bucket-key secrets.  

It also means that we assume the app should always have ownership rights of the bucket.   

Here's a <sup>contrived</sup>  case where I have 2 secrets for 2 different key pairs:
1. owner: full perms
2. web frontend: readonly

I would use the owner key pair secret to provision the claim.
Then manually bind the web frontend keypair to the bucket with a readOnly policy.  (We don't have plans yet for doing this in k8s)
Finally, I would mount the readonly key secret and the config map.

So storing the access keys and bucket data together reduces our flexibility.